### PR TITLE
Adding prompt & delayed fission in source code

### PR DESCRIPTION
**Pull request by mkreher13**
This PR includes prompt & delayed fission in nuclide.cpp (micro XS) and material.cpp (macro XS). These are essential changes for the transient capability that is being merged incrementally. 

Prompt & delayed nu_fission follow very similar code to nu_fission everywhere in these two C++ files. Delayed_nu_fission is a vector. I have expanded the xs_ array to allow room for prompt_nu_fission as well as 8 delayed precursor groups. This does not bar the use of fewer groups in the nuclear data.

To test the accuracy of this code: it is part of a fully functional transient code which has been successfully run on C5G7 benchmarks. When we get a little further in the incremental merge of that transient code, I do plan to include a test that would encompass many of these transient-related code additions. When the code is fully merged, I will be including documentation, examples, and tutorial notebooks. 

I have done a quick test to ensure that the calculation rate isn’t damaged by this addition. I ran the example in examples/pincell before and after including this code, and the calculation rate did not change noticeably: sometimes it was a little higher, and sometimes it was a little lower. So this addition did not slow down the code.

### CODE REVIEW

**Comment by mkreher13** [comment]
I can't explain why the triso regression test failed on just one system. It doesn't seem to have anything to do with the changes I've made. But let me know if I'm wrong.

File: include/openmc/particle.h
Comment (gridley) [inline include/openmc/particle.h:73]: Should probably use an `std::array<double, 8>` here since it's of a modest size and of known max size ahead of time. It will kill my GPU code performance to put an `std::vector` here. :smile:

File: include/openmc/particle.h
Comment (gridley) [inline include/openmc/particle.h:289]: Probably should have an underscore to match the style of surrounding variables.

File: src/material.cpp
Comment (gridley) [inline src/material.cpp:758]: It's slightly faster to do this instead:

```
  for (int d = 0; d < p.macro_xs_.delayed_nu_fission.size(); ++d) {
    p.macro_xs_.delayed_nu_fission[d] = 0.0;
 ```

File: src/material.cpp
Comment (gridley) [inline src/material.cpp:838]: Same as previous comment...

File: include/openmc/nuclide.h
Comment (gridley) [inline include/openmc/nuclide.h:130]: Pretty sure we should mark these as `static const int` since they are guaranteed to never change, and the compiler might optimize a little better if it knows that.

File: src/nuclide.cpp
Comment (gridley) [inline src/nuclide.cpp:395]: Same here on the loop style

File: src/nuclide.cpp
Comment (gridley) [inline src/nuclide.cpp:704]: It's possibly inefficient to call vector's `resize` in performance-critical code like this. I don't think in practice you'd see this because `vector`s tend to not let go of memory they've already allocated, at least with gcc's STL (although this is not guaranteed and may vary between C++ STL implementations AFAIK).

If you do end up sticking with `std::vector` in the micro XS cache, I recommend calling `resize` up front on each particle, in the constructor for this rather than here.

File: src/nuclide.cpp
Comment (gridley) [inline src/nuclide.cpp:714]: I recommend `std::fill` here...

**Comment by gridley** [comment]
Hey Miriam,

Just added some small comments on stuff that should be super quick to resolve. I didn't start a formal review since I think that's more a thing for one of the people labeled as a code maintainter. They'd probably have similar comments, I think.

Aside from that it's frickin' awesome to see some transient stuff finally getting put into OpenMC's public repo. Thinking about how we could easily do delayed precursor drift in MSRs now that this has been added... Nice.

File: src/material.cpp
Comment (gridley) [reply src/material.cpp:758]: On second thought, probably should just use `std::fill` here instead.

**Comment by gridley** [comment]
Another thing came to mind while I was cooking bacon this morning: you won't see much of a slowdown in the pin cell case since that one's not dominated by XS lookup. Because there is only one temperature, the XS caching efficiency is really high. I have XML input for a depleted fuel assembly with each pin at a different temperature, which causes the XS caching efficiency to plummet. This case is pretty much pure XS lookups in terms of run time.

 I think we should measure performance on this case before merging. Here are the zipped files. I'm fine to run it if you don't have time.

[depleted_different_temperatures_assembly.zip](https://github.com/openmc-dev/openmc/files/5721133/depleted_different_temperatures_assembly.zip)

File: include/openmc/nuclide.h
Comment (mkreher13) [reply include/openmc/nuclide.h:130]: I'm going to defer to the OpenMC team since I was just following precedent. I wanted to make XS_PROMPT_NU_FISSION and XS_DELAYED_NU_FISSION identical in form to XS_NU_FISSION. If there is a good case for modifying all these to static const int, let's do it in a separate pull request.

**Comment by mkreher13** [comment]
Thank you Gavin, great suggestions. I'll admit my experience in C++ is limited, so all these suggestions are helpful! 

I ran the example you sent, and this PR does not affect the calculation rate (I also ran it in my branch that has the full transient code - and it is slower there, so I might need your help to speed things up in the next few PRs I have planned).

File: include/openmc/nuclide.h
Comment (gridley) [reply include/openmc/nuclide.h:130]: Yeah, sounds good. The performance gain would be negligible anyways. I too dislike PR scope bloat, haha.

**Comment by gridley** [comment]
No prob Miriam! Thanks for making the changes. I'll leave this up a little bit longer for anyone else to provide input. And yeah, sounds fun to optimize the transient code some.

**Review by gridley — APPROVED** [review]

**Comment by paulromano** [comment]
Thanks for the PR @mkreher13 and the initial review @gridley. I'd like to review this before we merge so that I can run some performance tests as well but likely won't get to it until tomorrow at the earliest.

**Comment by gridley** [comment]
Oh, I was confusing something earlier. I just realized that the way this is currently written, it's going to drive memory usage up substantially since you're saving the pointwise (on the energy grid) nu-fission XS for each delayed neutron group. This IS where most of the memory is in OpenMC for smallish runs, IIRC, these `xs_` grids. And it will increase use 2.3x as much memory as it currently stands. I would appreciate not having that, particularly for GPU applications :smile: 

So I think the best approach is going to be indeed leaving separate entries for each of the delayed fission XS in the particle nuclide-wise micro XS cache, but not pre-calculating these quantities and storing them in `xs_`, instead multiplying the pre-tabulated fission XS by calling nu for each delayed group in each `calculate_xs` call. It will be a slight (pretty sure) performance hit, but less than half the memory usage. It shouldn't be much worse than looking up depletion reactions.

**Comment by mkreher13** [comment]
@gridley I think I understand your suggestion. I tried this in nuclide.cpp: 

I deleted lines 389-405, and then modified lines 695-703 to use micro.fission and nu instead of xs_. This looks like this:
micro.nu_fission = (1-f)*(micro.fission*nu(grid.energy[i_grid], EmissionMode::total)     etc...

Is that what you had in mind? This way we by-pass the storage of nu_fission, prompt_nu_fission, and delayed_nu_fission in xs_. This seems to work, but it must trigger computer precision because all the test results are then different just in the 6th or 7th decimal for example. That's annoying and I want to make sure I have it right before updating ALL the tests and revising the PR.

**Comment by gridley** [comment]
Eh, close, but not quite, I think. I was thinking: keep the normal nu fission as we've always had it so that stuff stays fast/exactly the same for the usual calculations, and don't add any extra room to the `xs_` array. Then, we'd need something like `settings::need_delayed_nu_fission` similar to `settings::need_depletion_rx` that could optionally trigger filling in the various delayed nu fission values in the NuclideMicroXS entries owned by each particle.

So you'd have something like this in particle.h:

```c++
struct NuclideMicroXS {
  // Microscopic cross sections in barns
  double total;            //!< total cross section
  double absorption;       //!< absorption (disappearance)
  double fission;          //!< fission
  double nu_fission;       //!< neutron production from fission

  double elastic;          //!< If sab_frac is not 1 or 0, then this value is
                           //!<   averaged over bound and non-bound nuclei
  double thermal;          //!< Bound thermal elastic & inelastic scattering
  double thermal_elastic;  //!< Bound thermal elastic scattering
  double photon_prod;      //!< microscopic photon production xs

  // Cross sections for depletion reactions (note that these are not stored in
  // macroscopic cache)
  double reaction[DEPLETION_RX.size()];

  // NEW THING HERE:
  // Nu fission cross sections for delayed neutron production
  double delayed_nu_fission[8]; // probably put in some constexpr variable where 8 is

  // Indicies and factors needed to compute cross sections from the data tables
  int index_grid;        //!< Index on nuclide energy grid
  int index_temp;        //!< Temperature index for nuclide
  double interp_factor;  //!< Interpolation factor on nuc. energy grid
  int index_sab {-1};    //!< Index in sab_tables
  int index_temp_sab;    //!< Temperature index for sab_tables
  double sab_frac;       //!< Fraction of atoms affected by S(a,b)
  bool use_ptable;       //!< In URR range with probability tables?

  // Energy and temperature last used to evaluate these cross sections.  If
  // these values have changed, then the cross sections must be re-evaluated.
  double last_E {0.0};      //!< Last evaluated energy
  double last_sqrtkT {0.0}; //!< Last temperature in sqrt(Boltzmann constant
                            //!< * temperature (eV))
};
```

Then leave the calculation of normal `nu_fission` as-is so that the normal flow of calculation would be uninterrupted, which I believe is causing the variation with your tests. The new code in nuclide.cpp would then look like:

```c++
    if (fissionable_) {
      // Calculate microscopic nuclide total cross section
      micro.fission = (1.0 - f)*xs(i_grid, XS_FISSION)
            + f*xs(i_grid + 1, XS_FISSION);

      // Calculate microscopic nuclide nu-fission cross section
      micro.nu_fission = (1.0 - f)*xs(i_grid, XS_NU_FISSION)
        + f*xs(i_grid + 1, XS_NU_FISSION);
    } else {
      micro.fission = 0.0;
      micro.nu_fission = 0.0;
    }

    // Calculate microscopic nuclide photon production cross section
    micro.photon_prod = (1.0 - f)*xs(i_grid, XS_PHOTON_PROD)
      + f*xs(i_grid + 1, XS_PHOTON_PROD);

   // --- NEW DELAYED NU SIG F CODE ---
   if (simulation::need_delayed_nu_fission) {
     for (int delayed_group=0; delayed_group<simulation::num_delayed_groups; ++delayed_group) {
       micro.delayed_nu_sig_f[delayed_group] = micro.fission * nu(grid.energy[i_grid], EmissionMode::delayed, delayed_group+1);
     }
   }

    // Depletion-related reactions
    if (simulation::need_depletion_rx) {
      // Initialize all reaction cross sections to zero
      for (double& xs_i : micro.reaction) {
        xs_i = 0.0;
      }
```

**Comment by mkreher13** [comment]
Ok, I see. I had a hunch that an if-statement surrounding the need for these cross sections would be helpful for performance. I have a whole settings entry lined up for "frequency_method_on". I'll go ahead and include that as soon as I can.

**Comment by mkreher13** [comment]
I converted this PR to a draft since I'm not ready for this to be merged, even if it was approved. I am looking at the repercussions of these suggested changes on the transient code, and it turns out they aren't trivial.

**Comment by gridley** [comment]
Dang, sorry to hear that Miram D:

@paulromano which do you think sounds better? Having OpenMC use 2.8x more memory in `xs_` arrays to precompute each delayed group's nu sigma fission, or recomputing the nu() reaction for delayed-groupwise nu fission on the fly? My preference is the latter.

**Comment by mkreher13** [comment]
No worries Gavin - I just needed more time to straighten out the larger transient code before committing to any of these suggestions. I successfully cleaned a lot up and this is already looking much better! I'm sure there are still performance enhancements to be made, and I'm open to hearing those and testing them out. 

This update introduces the Frequency Transform Method's unique settings which are used to generate if-statements specific to the FTM. So if the FTM is being used, micro cross sections for prompt & delayed nu fission will be calculated. If not, most of the code will remain unaltered. 

In my next PRs I will be using the FTM much more, so you will be seeing the overall functionality of these FTM settings soon.

**Comment by gridley** [comment]
Interesting, really looking forward to seeing where this goes!

File: include/openmc/particle.h
Comment (paulromano) [inline include/openmc/particle.h:73]:
```suggestion
  double prompt_nu_fission; //!< prompt neutron production from fission
  std::array<double, MAX_DELAYED_GROUPS> delayed_nu_fission; //!< delayed neutron production from fission
```

File: include/openmc/particle.h
Comment (paulromano) [inline include/openmc/particle.h:129]:
```suggestion
  double prompt_nu_fission; //!< macroscopic prompt production
  std::vector<double> delayed_nu_fission; //!< macroscopic delayed production
  double inverse_velocity; //!< macroscopic inverse velocity 
```

File: include/openmc/particle.h
Comment (paulromano) [inline include/openmc/particle.h:128]: Why isn't this `std::array` like in the micro?

File: openmc/settings.py
Comment (paulromano) [inline openmc/settings.py:79]: I would prefer to see these settings be added in a pull request where they are actually used rather than this one (where they don't appear to have any purpose).

File: src/nuclide.cpp
Comment (paulromano) [inline src/nuclide.cpp:710]: Why not evaluate nu at the interpolated energy and multiply by `micro.fission` rather than interpolating the product of nu and fission? Also, the indentation here looks inconsistent.

File: src/nuclide.cpp
Comment (paulromano) [inline src/nuclide.cpp:584]: Should this be inside `if (settings::frequency_method_on)`?

File: src/nuclide.cpp
Comment (paulromano) [inline src/nuclide.cpp:939]: Use `if (settings::frequency_method_on)`?

File: src/material.cpp
Comment (paulromano) [inline src/material.cpp:757]:
```suggestion
  std::fill(p.macro_xs_.delayed_nu_fission.begin(), p.macro_xs_.delayed_nu_fission.end(), 0.0);
```

File: src/material.cpp
Comment (paulromano) [inline src/material.cpp:837]:
```suggestion
        p.macro_xs_.delayed_nu_fission[d] += atom_density * micro.delayed_nu_fission[d];
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for this PR @mkreher13 and happy new year! Sorry it's taken a while for me to get around to reviewing it. I would prefer to restrict this PR to just the changes necessary to calculate the prompt/delayed nu fission values (and perhaps the `frequency_method_on` boolean). All of the other settings for the groups, frequencies, etc. would belong better in a PR where they are actually used.

File: include/openmc/particle.h
Comment (mkreher13) [reply include/openmc/particle.h]: When I use p.macro_xs_.delayed_nu_fission later in src/mgxs.cpp I have to resize the vector to exactly the number of actual delayed groups to avoid errors. An array, being fixed size, doesn't have that capability from what I read. 
When I submit those changes to mgxs.cpp in a future PR, if you see a workaround, I am happy to modify it from a vector to an array. But for now, I don't see a way to make it an array.

File: src/nuclide.cpp
Comment (mkreher13) [reply src/nuclide.cpp:710]: The interpolation of a product is not equal to the interpolation of the parts of that product. I think it would introduce more error to interpolate them separately.
Indentation: sorry, it looks fine in my editor and on my repository in GitHub, but I agree it is wonky here in the PR. Not sure why.

File: openmc/settings.py
Comment (mkreher13) [reply openmc/settings.py]: I agree with you! They would fit better in a separate PR so I took them out. I did however leave the frequency_mesh since the presence of the frequency_mesh is necessary to turn frequency_method_on to True.

File: include/openmc/particle.h
Comment (paulromano) [reply include/openmc/particle.h]: That makes sense -- we can leave it as `vector`

File: src/nuclide.cpp
Comment (paulromano) [reply src/nuclide.cpp:712]: > The interpolation of a product is not equal to the interpolation of the parts of that product.

Understood, but it's actually not clear which will be correct. Also, sometimes `nu` is given as a polynomial function rather than a tabulated function, in which case it doesn't make sense to interpolate it when you can evaluate it exactly at the energy of interest. Performance wise, you're also better off evaluating nu once rather than twice.

Regarding indentation -- it looks like you have tab characters in here. We uniformly use spaces in the OpenMC code base, so I would encourage you to ensure that your editor converts tabs to spaces when working on OpenMC.

**Comment by gridley** [comment]
As a follow on, does `frequency_transform_on` imply a transient calculation and a transient calculation imply `frequency_transform_on`?

**Comment by mkreher13** [comment]
@gridley not exactly. The frequency_method_on implies a frequency transformed transient. There is also the capability to run an adiabatic transient. In that case, there is no need to separate out prompt & delayed fission since there are no precursor frequencies to multiply the delayed fission with. Actually, the adiabatic transient requires no changes to the source code, only the presence of frequencies requires a modification to how things are sampled.

**Comment by gridley** [comment]
Ah, I didn't know you were planning to add adiabatic transients as well. That will be nice to compare against. Although IIRC isn't adiabatic like, really bad accuracy typically?

File: src/nuclide.cpp
Comment (mkreher13) [reply src/nuclide.cpp]: I may have misunderstood your first comment. I'm on board with what you're proposing, and it's in the new commit. I also made sure spaces are being used. So hopefully that takes care of the indentations.

Comment (mkreher13) [reply src/nuclide.cpp]: I may have misunderstood your first comment. I'm on board with what you're proposing, and it's in the new commit. I also made sure spaces are being used. So hopefully that takes care of the indentations.

File: src/settings.cpp
Comment (paulromano) [inline src/settings.cpp:551]:
```suggestion
```

File: src/material.cpp
Comment (paulromano) [inline src/material.cpp:756]: I'm a little confused -- I thought you said this needs to be sized to the actual number of precursor groups, but here the size is effectively static.

**Review by paulromano — CHANGES_REQUESTED** [review]
Can you merge the develop branch into this? That should fix the issue with the expected CI jobs not showing up.

**Comment by paulromano** [comment]
I did some performance testing on a single assembly problem with fresh and depleted fuel compositions. With this branch, performance drops about 4% for the fresh fuel version (both with or without windowed multipole active) and about 2% for the depleted fuel version. I'm not sure there's much we can do about that though given that adding more members to those cross section cache structs is expected to hurt performance. I'll keep thinking about whether there's a way to avoid the performance hit.

**Comment by gridley** [comment]
Might be ugly, but maybe there could be some templates to handle this. This would even be beneficial for dropping the depletion reactions slots in that struct out, which of course won't be used in any transient and aren't used in a lot of OpenMC calculations as is.

So i think you'd have a particle class templated on the micro XS cache type, and correspondingly make some class called `Simulation` or something like that that fills out these templates. So there could be `Simulation<RunMode::transient>`, `Simulation<RunMode::depletion>`, `Simulation<RunMode::eigenvalue` or something along those lines. Similarly, the fact that we have photons/electrons/protons also reserving space for depletion reactions is crazy. Definitely some improvements to be made here.

Man, this is where C++ 20 concepts would be really nice. A particle concept would probably be a clean way to implement these sorts of specializations. Totally get that we're not going to be able to use those for a little bit though.

File: src/material.cpp
Comment (mkreher13) [reply src/material.cpp:756]: It's in a different part of the code, not included in this PR. When I use p.macro_xs_.delayed_nu_fission later in src/mgxs.cpp I have to resize the vector to exactly the number of actual delayed groups to avoid errors.

**Comment by mkreher13** [comment]
It's not surprising that new features cost performance. The question is what can be done to minimize it? What performance drop is the limit? How will future new features be handled? I think Gavin has a good point: OpenMC may eventually have to move to a different structure where features can be selected so that the most basic use has the fastest performance. 

I am continually open to re-formatting the transient code to be the most effective as possible, so I'm really grateful for all the comments on this PR to do just that! Also, once it is fully merged and there are tests, we can really re-evaluate the whole structure and re-organize for performance (a little hard to do that right now without the whole feature in place).

File: src/material.cpp
Comment (gridley) [reply src/material.cpp:756]: Either way, people studying transients know immediately where the 8 comes from, but it would be good to define a `constexpr` variable like `MAX_PRECURSOR_GROUPS` rather than using the literal number 8 here. Just makes it more clear if some non-transient-experienced person has to work on this.

**Review by gridley — COMMENTED** [review]
I think it would be good to get rid of the literal number `8` here and replace it with a compile-time variable for better clarity. Also, I am somewhat opposed to merging partial implementations of features that are going to slow down the code a little bit, especially since the newly introduced code isn't really tested yet to be correct. OTOH you can work early feedback in this way. Hm. Definitely pros and cons to each approach.

File: src/material.cpp
Comment (mkreher13) [reply src/material.cpp:756]: I can easily use MAX_DELAYED_GROUPS here instead, no problem.

**Comment by mkreher13** [comment]
That will have to be a decision for @paulromano since he has previously asked me to make this merge as incremental as possible. I have the full code implemented in a different branch and could easily do a massive PR to get it in all at once. What I'm seeing with this PR is that the incremental approach allows much more focused comments and improvements along the way. 

As for testing, we can certainly add tests along the way. Here for instance, to make sure the sum of all the delayed + prompt = total or something like that. I have done this test myself in the implementation, but I could try my hand at making a unit test.

**Comment by gridley** [comment]
Well, yeah, of Paul wants it, then he's the experienced one here to listen to. Come to think of it, given the way OpenMC has periodic releases, I guess it does make sense to incrementally build in this way.

**Comment by paulromano** [comment]
There are a few reasons I encourage people to shoot for multiple smaller PRs rather than one huge PR. For one, it makes the burden of review easier, allow the reviewer to focus on one aspect at a time. It also encourages contributors to ensure their branch doesn't stray too far from the develop branch, which can cause problems if they diverge from one another. Also, by seeking feedback from the development team earlier, it can help us identify whether a large planned feature will be considered — this doesn't happen very frequently (and I don't expect this to be the case for your work), but I would hate for someone to work on a huge feature only to discover at the end of the day that it wouldn't meet the standards for inclusion in the develop branch.

That being said, the small PR philosophy depends on PRs being self-contained. After thinking this over, I think the current PR doesn't really make sense as a standalone contribution because the prompt/delayed cross sections being calculated are not actually being used anywhere, nor is the mesh specification. I'm going to recommend that we table this PR and that it be included as part of larger PR where the features that are added are a necessary part of the design. Sorry for the churn on this @mkreher13!

**Comment by mkreher13** [comment]
Makes sense! Ben really wants this merged, so I'll try to do it right to maximize our chances. 

I'll close this PR and open a new one focused solely on the adiabatic method. That will give you a chance to explore the kinetics code without any modifications to the source code. So once I have a test & an example ready, I'll initiate that.

**Pull request closed without merging**
